controllers/etcd_controller.go

```diff
@@ -963,33 +961,53 @@ func (r *EtcdReconciler) updateEtcdErrorStatus(etcd *druidv1alpha1.Etcd, lastErr
 }
 
 func (r *EtcdReconciler) updateEtcdStatus(etcd *druidv1alpha1.Etcd, svc *corev1.Service, ss *appsv1.StatefulSet) error {
+
+	sts, err := r.waitUntilStatefulSetReady(ss)
+	if err != nil {
+		return err
+	}
+
 	svcName := svc.Name
 	etcd.Status.Etcd = druidv1alpha1.CrossVersionObjectReference{
-		APIVersion: ss.APIVersion,
-		Kind:       ss.Kind,
-		Name:       ss.Name,
+		APIVersion: sts.APIVersion,
+		Kind:       sts.Kind,
+		Name:       sts.Name,
 	}
 	conditions := []druidv1alpha1.Condition{}
-	for _, condition := range ss.Status.Conditions {
+	for _, condition := range sts.Status.Conditions {
 		conditions = append(conditions, convertConditionsToEtcd(&condition))
 	}
 	etcd.Status.Conditions = conditions
 
 	// To be changed once we have multiple replicas.
-	etcd.Status.CurrentReplicas = ss.Status.CurrentReplicas
-	etcd.Status.ReadyReplicas = ss.Status.ReadyReplicas
-	etcd.Status.UpdatedReplicas = ss.Status.UpdatedReplicas
-	etcd.Status.Ready = (health.CheckStatefulSet(ss) == nil)
+	etcd.Status.CurrentReplicas = sts.Status.CurrentReplicas
+	etcd.Status.ReadyReplicas = sts.Status.ReadyReplicas
+	etcd.Status.UpdatedReplicas = sts.Status.UpdatedReplicas
 	etcd.Status.ServiceName = &svcName
 	etcd.Status.LastError = nil
 	etcd.Status.ObservedGeneration = &etcd.Generation
+	etcd.Status.Ready = true
 
 	if err := r.Status().Update(context.TODO(), etcd); err != nil && !errors.IsNotFound(err) {
 		return err
 	}
 	return r.removeOperationAnnotation(etcd)
 }
 
+func (r *EtcdReconciler) waitUntilStatefulSetReady(sts *appsv1.StatefulSet) (*appsv1.StatefulSet, error) {
+	ss := &appsv1.StatefulSet{}
+	err := gardenerretry.UntilTimeout(context.TODO(), DefaultInterval, DefaultTimeout, func(ctx context.Context) (bool, error) {
+		if err := r.Get(context.TODO(), types.NamespacedName{Name: sts.Name, Namespace: sts.Namespace}, ss); err != nil {
```

Comment:
Shouldn't the controller create the `StatefulSet` if it is not found? Can we not shuffle this logic with the overall reconciliation logic to handle this smoothly?

Reply:
The controller creates/reconciles the statefulset. Here, it checks if the statefulset is ready. it can be because of misconfiguration of etcd resource resulting in pods failing(eg. bucket name provided wrongly).

Reply:
Yes. Why not do this check where the controller reconciles the statefulset?

Reply:
Yup. Makes sense. Ill move the logic into the reconcile cycle before updating the status.